profiling/src/lib.rs

```diff
@@ -0,0 +1,713 @@
+        let handle = module.handle;
+        module.handle = std::ptr::null_mut();
+        handle
+    };
+
+    let extension = ZendExtension {
+        name: NAME.as_ptr(),
+        version: VERSION.as_ptr(),
+        author: b"Datadog\0".as_ptr(),
+        url: b"https://github.com/DataDog\0".as_ptr(),
+        copyright: b"Copyright Datadog\0".as_ptr(),
+        startup: Some(startup),
+        shutdown: Some(shutdown),
+        ..Default::default()
+    };
+
+    static mut EXTENSION_ENTRY: MaybeUninit<ZendExtension> = MaybeUninit::uninit();
+    EXTENSION_ENTRY.write(extension);
+
+    PREV_INTERRUPT_FUNCTION.write(zend::zend_interrupt_function);
+    zend::zend_interrupt_function = Some(if zend::zend_interrupt_function.is_some() {
+        interrupt_function_wrapper
+    } else {
+        interrupt_function
+    });
+
+    PREV_EXECUTE_INTERNAL.write(zend::zend_execute_internal.unwrap_or(zend::execute_internal));
+    zend::zend_execute_internal = Some(execute_internal);
+
+    zend::zend_register_extension(&mut *EXTENSION_ENTRY.as_mut_ptr(), handle);
+
+    // Set up runtime-id for Code Hotspots
+    RUNTIME_ID = profiling::Uuid::from(uuid::Uuid::new_v4());
+
+    ZendResult::Success
+}
+
+extern "C" fn ginit(global: *mut c_void) {
+    #[cfg(debug_assertions)]
+    {
+        logging::log_init(LevelFilter::Trace);
+        trace!("GINIT({:p})", global);
+    }
+
+    // Safety: The engine has allocated the globals to the right size and
+    // alignment, so it is safe to dereference this, but only when cast to
+    // MaybeUninit, as it's uninitialized (which is the point of this hook).
+    unsafe {
+        let global = global as *mut MaybeUninit<DatadogPhpProfilingGlobals>;
+        (*global).write(DatadogPhpProfilingGlobals::default());
+    }
+}
+
+unsafe extern "C" fn gshutdown(_global: *mut c_void) {
+    #[cfg(debug_assertions)]
+    trace!("GSHUTDOWN({:p})", _global);
+}
+
+extern "C" fn prshutdown() -> ZendResult {
+    #[cfg(debug_assertions)]
+    trace!("PRSHUTDOWN");
+
+    ZendResult::Success
+}
+
+/// The version of PHP at runtime, not the version compiled against. Sent as
+/// a profile tag.
+static mut PHP_VERSION: Option<String> = None;
+
+static mut PROFILER: RefCell<Option<Profiler>> = RefCell::new(None);
+
+/// The runtime ID, which is basically a universally unique "pid", so it can
+/// change on minit, as well as theoretically on fork. It should only be
+/// changed from thread-safe contexts.
+/// todo: support forking.
+static mut RUNTIME_ID: profiling::Uuid = profiling::Uuid::new();
+
+/// The version of the ZendEngine at runtime, not the version compiled against.
+/// It's currently unused, but I hope to send it as a metric or something soon
+/// so I'm keeping it here so I don't have to look up how to get it again.
+static mut ZEND_VERSION: Option<String> = None;
+
+/// # Safety
+/// Only call during an active request from the PHP thread!
+unsafe fn getenv(name: &CStr) -> Option<String> {
```

Comment:
Why does this take a `&CStr`?  You can simplify both here and at the callsites by taking an `&str`(guaranteeing UTF8-ness at compile-time) and before falling back to libc, allocating a temporary CString to get the null terminator (which is basically free since all of the environment variable names you use are a dozen bytes or so).